finetune/network_modules.py

```diff
@@ -350,3 +351,104 @@ def seq_lab_internal(hidden):
                 'transition_matrix': transition_params
             }
         }
+
+
+def association(hidden, pool_idx, targets, n_targets, config, train=False, reuse=None, **kwargs):
+    """
+    An Attention based sequence labeler model with association.
+
+    :param hidden: The output of the featurizer. [batch_size, sequence_length, embed_dim]
+    :param pool_idx: the index of the classify tokens along the sequence dimension. [batch_size]
+    :param targets: A dict containing:
+     'labels': The sequence labeling targets. [batch_size, sequence_length],
+     'associations': A matrix of class ids for the associations [batch_size, sequence_length, seqence_length]
+    :param n_targets: A python int containing the number of classes that the model should be learning to predict over.
+    :param config: A config object, containing all parameters for the featurizer.
+    :param train: If this flag is true, dropout and losses are added to the graph.
+    :param reuse: Should reuse be set within this scope.
+    :param kwargs: Spare arguments.
+    :return: dict containing:
+        "logits": The un-normalised log probabilities of each class being in each location. For usable predictions,
+            sampling from this distrobution is not sufficiant and a viterbi decoding method should be used.
+        "losses": The negative log likelihood for the sequence targets.
+        "predict_params": A dictionary of params to be fed to the viterbi decode function.
+    """
+    with tf.variable_scope('sequence-labeler', reuse=reuse):
+        nx = config.n_embed
+        length = config.max_length
+        num_associations = len(config.possible_associations) + 1
+
+        def seq_lab_internal(hidden):
+            attn_fn = functools.partial(
+                attn, scope="seq_label_attn", n_state=nx, n_head=config.seq_num_heads,
+                resid_pdrop=config.resid_p_drop, attn_pdrop=config.attn_p_drop,
+                train=train, scale=False, mask=False
+            )
+            n = norm(attn_fn(hidden) + hidden, 'seq_label_residual')
+            flat_logits = tf.layers.dense(n, n_targets)
+            logits = tf.reshape(flat_logits, tf.concat([tf.shape(hidden)[:2], [n_targets]], 0))
+
+            association_head = tf.layers.dense(n, nx)
+            association_head = tf.reshape(association_head, tf.concat([tf.shape(hidden)[:2], [nx]], 0))
+
+            a = tf.expand_dims(association_head, 1)
+            b = tf.expand_dims(association_head, 2)
+
+            probs = tf.nn.softmax(logits)
+
+            features = tf.concat(
+                [
+                    a - b, a * b,
+                    tf.tile(a, [1, length, 1, 1]),
+                    tf.tile(b, [1, 1, length, 1]),
+                    tf.tile(tf.expand_dims(probs, 1),[1, length, 1, 1]),
```

Comment:
Hrmmm ok.... feels weird to directly pass the output of the softmax as features.  I agree we want this information to be accessible by the association head, but think it might warrant some more thought about how to make sure that happens.